package com.test.tcqt.core.env

import android.app.Application
import android.content.Context
import com.test.tcqt.core.log.Log
import com.tencent.mobileqq.vas.theme.api.ThemeUtil

internal object HookEnv {

    const val TIM_PACKAGE = "com.tencent.tim"
    const val QQ_PACKAGE = "com.tencent.mobileqq"

    val moduleClassLoader: ClassLoader by lazy { this::class.java.classLoader!! }

    lateinit var hostAppPackageName: String
        private set

    lateinit var processName: String
        private set

    lateinit var moduleApkPath: String
        private set

    lateinit var hostApkPath: String
        private set

    lateinit var appName: String
        private set

    lateinit var versionName: String
        private set

    var versionCode: Long = -1L
        private set

    lateinit var hostAppContext: Context
        private set

    lateinit var application: Application
        private set

    lateinit var hostClassLoader: ClassLoader
        private set

    lateinit var moduleDataPath: String
        private set

    fun setApplication(app: Application) {
        application = app
    }

    fun setHostClassLoader(cl: ClassLoader) {
        hostClassLoader = cl
    }

    fun setHostApkPath(path: String) {
        hostApkPath = path
    }

    fun setVersionName(v: String) {
        versionName = v
    }

    fun setVersionCode(code: Long) {
        versionCode = code
    }

    fun setHostAppContext(ctx: Context) {
        hostAppContext = ctx
    }

    fun setModuleApkPath(path: String) {
        moduleApkPath = path
    }

    fun setProcessName(p: String) {
        processName = p
    }

    fun setHostAppPackageName(pkg: String) {
        hostAppPackageName = pkg
    }

    fun setAppName(name: String) {
        appName = name
    }

    fun setModuleDataPath(path: String) {
        moduleDataPath = path
    }

    fun isTIM() = hostAppPackageName == TIM_PACKAGE

    fun isQQ() = hostAppPackageName == QQ_PACKAGE

    fun isNT() = try {
        load("com.tencent.qqnt.base.BaseActivity") != null
    } catch (_: Exception) {
        false
    }

    fun isMainProcess() = ::processName.isInitialized &&
            ::hostAppPackageName.isInitialized &&
            processName == hostAppPackageName

    fun requireMinQQVersion(versionCode: Long): Boolean {
        return this.isQQ() && this.versionCode >= versionCode
    }

    fun requireMinTimVersion(versionCode: Long): Boolean {
        return this.isTIM() && this.versionCode >= versionCode
    }

    fun String.toHostClass(): Class<*> = loadOrThrow(this)

    fun String.toHostClassOrNull(): Class<*>? = load(this).also {
        if (it == null) {
            Log.e("class: $this not found")
        }
    }

    fun isNightMode(): Boolean {
        return ThemeUtil.isNowThemeIsNight(null, true, null) ||
                isQQ() &&
                requireMinQQVersion(QQVersion.QQ_9_2_55_BETA_32895) &&
                ThemeUtil.isThemeNightModeV2()
    }

    fun resetApp() {
        PlatformTools.killSubProcesses()
        PlatformTools.reStartLoadingActivity()
    }
}
